src/rgw/rgw_sal_filter.h

```diff
@@ -16,308 +16,544 @@
 #pragma once
 
 #include "rgw_sal.h"
+#include "rgw_oidc_provider.h"
+#include "rgw_role.h"
 
 namespace rgw { namespace sal {
 
+class FilterCompletions : public Completions {
+protected:
+  std::unique_ptr<Completions> next;
+
+public:
+  FilterCompletions(std::unique_ptr<Completions> _next) : next(std::move(_next)) {}
+  virtual ~FilterCompletions() = default;
+  virtual int drain() override { return next->drain(); }
+};
+
+class FilterPlacementTier : public PlacementTier {
+protected:
+  std::unique_ptr<PlacementTier> next;
+
+public:
+  FilterPlacementTier(std::unique_ptr<PlacementTier> _next) : next(std::move(_next)) {}
+  virtual ~FilterPlacementTier() = default;
+
+  virtual const std::string& get_tier_type() override { return next->get_tier_type(); }
+  virtual const std::string& get_storage_class() override { return next->get_storage_class(); }
+  virtual bool retain_head_object() override { return next->retain_head_object(); }
+
+  /* Internal to Filters */
+  PlacementTier* get_next() { return next.get(); }
```

Comment:
are these `get_next()` functions used anywhere? it might not be a good idea to expose these directly, because a caller could use them to circumvent a filter's behavior

Reply:
They are used internally, and it's expected that filters derived from this code will use them as will.  D4NFilter certainly does.

Reply:
derived filters already have access to the protected `next` variable. if you want to provide a getter for that, can it also be `protected`?

Reply:
So, for example, FilterMultipartUpload::complete needs access to FilterObject::next

Reply:
i see, thanks. i would use friend declarations to preserve the encapsulation here, but it's up to you